File: LAB3_inClass/Ex1/app/src/main/java/com/example/ex1/MainActivity.java
package com.example.ex1;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.ListView;
import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    ListView listView;
    ArrayList<User> arrayUser;
    UserAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        listView = findViewById(R.id.myListView);
        arrayUser = new ArrayList<>();

        // Thêm 20 người dùng vào danh sách
        arrayUser.add(new User("Trương Mạnh Nguyên", "Nha Trang", R.mipmap.ic_launcher));
        arrayUser.add(new User("Harry", "San Diego", R.mipmap.ic_launcher));
        arrayUser.add(new User("Marla", "San Francisco", R.mipmap.ic_launcher));
        arrayUser.add(new User("Sarah", "San Marco", R.mipmap.ic_launcher));
        arrayUser.add(new User("Zhongli", "Liyue", R.mipmap.ic_launcher));
        arrayUser.add(new User("Venti", "Mondstadt", R.mipmap.ic_launcher));
        arrayUser.add(new User("Raiden", "Inazuma", R.mipmap.ic_launcher));
        arrayUser.add(new User("Nahida", "Sumeru", R.mipmap.ic_launcher));
        arrayUser.add(new User("Furina", "Fontaine", R.mipmap.ic_launcher));
        arrayUser.add(new User("Mavuika", "Natlan", R.mipmap.ic_launcher));
        arrayUser.add(new User("Sinh viên A", "VNU-HCM", R.mipmap.ic_launcher));
        arrayUser.add(new User("Sinh viên B", "UIT", R.mipmap.ic_launcher));
        arrayUser.add(new User("David", "London", R.mipmap.ic_launcher));
        arrayUser.add(new User("Emma", "Paris", R.mipmap.ic_launcher));
        arrayUser.add(new User("Lisa", "Tokyo", R.mipmap.ic_launcher));
        arrayUser.add(new User("Kevin", "Seoul", R.mipmap.ic_launcher));
        arrayUser.add(new User("Anna", "Berlin", R.mipmap.ic_launcher));
        arrayUser.add(new User("Peter", "Sydney", R.mipmap.ic_launcher));
        arrayUser.add(new User("Laura", "Toronto", R.mipmap.ic_launcher));
        arrayUser.add(new User("Tom", "New York", R.mipmap.ic_launcher));

        adapter = new UserAdapter(this, R.layout.item_user, arrayUser);
        listView.setAdapter(adapter);
    }
}

File: LAB3/Exercise1/app/src/main/java/com/example/exercise/MainActivity.kt
package com.example.exercise

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    // Khai báo các biến sẽ sử dụng
    private lateinit var viewModel: NewsViewModel
    private lateinit var adapter: NewsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // 1. Tìm cái RecyclerView mà đã thiết kế trong file activity_main.xml
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)

        // 2. Quy định cách sắp xếp: Danh sách cuộn dọc từ trên xuống dưới
        recyclerView.layoutManager = LinearLayoutManager(this)

        // 3. Khởi tạo Adapter
        adapter = NewsAdapter()
        recyclerView.adapter = adapter

        // 4. Gọi ViewModel
        viewModel = ViewModelProvider(this)[NewsViewModel::class.java]

        // 5. Mở một luồng chạy ngầm để liên tục lắng nghe dữ liệu từ mạng tải về
        lifecycleScope.launch {
            // collectLatest: Bất cứ khi nào tải xong 1 trang báo mới, nó sẽ báo về đây
            viewModel.articles.collectLatest { pagingData ->
                // Nạp dữ liệu vừa tải được vào Adapter để vẽ lên màn hình
                adapter.submitData(pagingData)
            }
        }
    }
}